[TECH] Remplacer le package redis par ioredis.

VincentHardouin opened the pull request:
## :unicorn: Problème
Nous sommes en retard sur la version du package `redis`. En effet, [la version 4 est disponible depuis novembre 2021](https://github.com/redis/node-redis/blob/master/CHANGELOG.md#v400---24-nov-2021). 
Cette version comporte plusieurs changements majeurs : 
- Les fonctions sont désormais des promesses 🚀 
- Le nommage des fonctions est désormais en camelCase 
- Il est obligatoire de se connecter avant d'effectuer une action `await redisClient.connect();`

En faisant la migration vers cette version, nous nous sommes rendu compte que `Redlock` ne fonctionnait plus, ce dernier utilise la fonction `evalsha` proposée par le client Redis. Comme on a pu le voir cette fonction a été renommée en camelCase par `evalSha`. En investiguant du côté de Redlock, [ce dernier dit être prévu pour ioredis](https://github.com/mike-marcacci/node-redlock#configuration) dans sa future version et [qu'il faut utiliser `redis` avec le `legacyMode`](https://github.com/mike-marcacci/node-redlock/issues/163) (ce qui rétablit l'ancien nommage). Nous n'envisageons pas cette solution. 

[`ioredis`](https://github.com/redis/ioredis) et [`redis`](https://github.com/redis/node-redis) sont tous les deux des clients Redis compris dans l'organisation Github `redis` et [ils ont un nombre comparable de téléchargements par semaine ](https://npmtrends.com/ioredis-vs-redis) : 

<img width="1296" alt="Screenshot 2023-07-25 at 16 42 02" src="https://github.com/1024pix/pix/assets/26384707/1fe9eaaa-4ff0-4b51-94cb-908cbdc7af55">


## :robot: Proposition
Nous avons donc décidé de passer à ioredis. 
Pour effectuer cette migration de manière sereine, nous avons ajouté des tests sur les méthodes que la classe `RedisClient` propose. 
Une fois ioredis installée, comme celui-ci supporte nativement les promesses, nous avons alors pu retirer les `promisify`.

## :rainbow: Remarques

### Fermeture multiple de la même connexion 

#### Cache 
Une fois des tests manuels effectués, nous sommes rendus compte que le `learningContentCache` fermait plusieurs fois la même connexion à Redis, nous avons corrigé cela en appelant `underlyingCache` qui se charge lui-même d'appeler les autres caches. 

#### Ctrl-C
Nous avons constaté un souci connu lors d'un `Ctrl-C` depuis une commande npm qui déclenche deux `SIGINT`. Le `Ctrl-C` va déclencher un SIGINT au process npm et Node, puis le process npm va relayé le signal au process Node qu'il gère. Le process Node.js reçoit donc deux fois le signal. Dans notre cas cela déclenche deux fois la fermeture aux connexions Redis et lève donc des erreurs, car celles-ci sont déjà fermées. 
Si cela vous intéresse [cet article en parle](https://lisk.com/blog/posts/why-we-stopped-using-npm-start-child-processes)

Pour gérer ce cas, nous avons fait en sorte de vérifier si la connexion n'était pas déjà fermée avant de la fermer. Ce qui évite de throw une erreur inutilement. 

### Fonctionnement du lock Redis
Au passage, en écrivant un test pour la méthode `lockDisposer`, cela nous a permis de voir comment un lock Redis fonctionne. Comme nous avons qu'un Redis, le fonctionnement est assez simple : 

- La connexion qui souhaite mettre à jour une clé vérifie qu'un lock n'est pas en place, plus concrètement utilise la commande `EXISTS <key>` pour vérifier que la clé n'est pas présente 
- Si le lock n'est pas présent, alors il en créé un, c'est-à-dire enregistre un couple clé/valeur avec un TTL `SET <key> <value> PX <ttl-in-milliseconds>` 
- Le lock est alors en place, soit il est détruit à cause du TTL, soit parce qu'on le retire `DEL <key>`. 

Pour en savoir plus [la doc Redis est bien.](https://redis.io/docs/manual/patterns/distributed-locks/)

Ce qui nous a perturbés c'est qu'il n'y a pas de corrélation entre le nom du lock et la clé que nous souhaitons modifier, dans l'implémentation de notre cache, le nom du lock correspond à `locks:<nom-de-la-clé-que-nous-allons-modifier>`

### Fonctionnement du disposer

https://github.com/1024pix/pix/blob/77dac7bd8352a87e48461f3e23b39a4709175426/api/lib/infrastructure/caches/RedisCache.js#L41-L42

Le lockDisposer est le code à exécuter à la fin dans tous les cas, on pourrait comparer ce pattern à un try/catch/finally : 
```javascript
try {
	retrieveAndSetValue()
} catch (e) {
} finally {
	lockDisposer()
}
```

[Le code source de la méthode `using` de Bluebird](https://github.com/petkaantonov/bluebird/blob/a9c04913b5e6d76f3d616b393b9f1e19985c7193/src/using.js) confirme cela : 

https://github.com/petkaantonov/bluebird/blob/a9c04913b5e6d76f3d616b393b9f1e19985c7193/src/using.js?plain=0#L192-L195

```javascript
		var promise = resultPromise.lastly(function() {
            var inspection = new Promise.PromiseInspection(resultPromise);
            return dispose(resources, inspection);
        });
```

On peut voir un [lastly qui est un alias pour finally](http://bluebirdjs.com/docs/api/finally.html)

## :100: Pour tester

- Se connecter à la RA et constater le bon fonctionnement du référentiel 
- Vérifier le bon rafraîchissement de token 
- Lancer sur la RA ou en local `npm run cache:refresh` et constater le bon fonctionnement de la commande.

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr6695.review.pix.fr)
  * [App (.org)](https://app-pr6695.review.pix.org)
  * [Orga (.fr)](https://orga-pr6695.review.pix.fr)
  * [Orga (.org)](https://orga-pr6695.review.pix.org)
  * [Certif (.fr)](https://certif-pr6695.review.pix.fr)
  * [Certif (.org)](https://certif-pr6695.review.pix.org)
  * [Admin](https://admin-pr6695.review.pix.fr)
  * [API](https://api-pr6695.review.pix.fr/api/)
  * [Pix1D](https://1d-pr6695.review.pix.fr)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr6695/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr6695/environment)

VincentHardouin committed:
sr(api): add missing tests on RedisClient

Co-authored-by: Alexandre Coin <alexandre.coin@pix.fr>

VincentHardouin committed:
chore(api): migrate from redis to ioredis

Co-authored-by: Alexandre Coin <alexandre.coin@pix.fr>

VincentHardouin committed:
sr(api): remove useless redis package

Co-authored-by: Alexandre Coin <alexandre.coin@pix.fr>

VincentHardouin committed:
chore(api): remove redundant disconnection to redis

Underlying cache already quit redis cache and distributed cache

Co-authored-by: Alexandre Coin <alexandre.pc.coin@gmail.com>

VincentHardouin committed:
feat(api): handle multiple SIGINT in redis quit method

SIGINT signal can be sent multiple times, but we can't quit redis connection several times. We should handle this to avoid error

francois2metz reviewed (approved):
:rose: :hibiscus: :white_flower: :wilted_flower:  :tulip: :sunflower: :cherry_blossom:  LGTM :cherry_blossom: :sunflower:  :tulip: :wilted_flower: :white_flower: :hibiscus: :rose:
